### PR TITLE
Parametric varchar v4

### PR DESCRIPTION
**Pull request by pnowojski**
Supersedes https://github.com/facebook/presto/pull/4311

### CODE REVIEW

**Comment by pnowojski** [comment]
I have reverted accidental `push --force` to this pull request

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -4974,13 +4977,13 @@ public void testInvalidType()
         computeActual("SELECT CAST(null AS array<foo>)");
     }
 
-    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "\\Qline 1:21: '+' cannot be applied to varchar, bigint\\E")
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "\\Qline 1:21: '+' cannot be applied to varchar(2147483647), bigint\\E")
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Why is the type of `'a' || 'z'` VARCHAR(MAX)? It should be VARCHAR(2), no?

**Comment by martint** [comment]
Other than how we avoid breaking backwards compatibility and minor comments, this looks pretty good. I assume it passes all integration tests, but I'll try running it through our verifier to make sure nothing breaks.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 1 above)
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Currently variable argument functions do not play well with arguments using types parametrized.
The variable argument functions mechanism requires that all the arguments are of exactly same type. We cannot enforce that base must match but parameters are the same.

What could possible be done to change definition of `concat`

Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Sounds good. Can you add a comment in this test with that explanation? I have the feeling we've already discussed this and I forgot about it :)

**Comment by losipiuk** [comment]
Superseded by https://github.com/facebook/presto/pull/4420

**Comment by losipiuk** [comment]
I confirm integration tests are passing. 
I added some comments on backward compatibility. Let me know what you think.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 1 above)
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: done

**Pull request closed without merging**